### ISSUE TITLE
Issue with doing transaction for ERC721 precompile.

### ISSUE THREAD

**Issue by KarishmaBothara** [cennznet/cennznet#709]
Steps

1. Start local node. Get your metamask's first address. To find the equivalent cennznet address of a evm address, search with you evm address on https://nikau.uncoverexplorer.com/, it will give you the cennznet address. Send some cpay from Alice to your eth address (https://cennznet.io/#/accounts) connect to local node
2. Switch Metamask to use local node (2999 for chainId, http://localhost:9933/ for url) - add network.
3. Create a NFT with eth-wallet, can use the app-hub branch below to create collection, series and set name commenting/uncommenting the tx required (https://github.com/cennznet/app-hub/blob/spike/mintNFTFromEthWallet/pages/_app.tsx#L99)
https://github.com/cennznet/app-hub/tree/spike/mintNFTFromEthWallet
4. Once created this will be first NFT on local node at collection 0 and series 0 so the precompile address is - 0xaAAAaaAA00000000000000000000000000000000
5. Use the smart contract as below

```
pragma solidity ^0.8.13;
import "@openzeppelin/contracts/interfaces/IERC721.sol";
import "@openzeppelin/contracts/interfaces/IERC165.sol";
import "@openzeppelin/contracts/token/ERC721/extensions/IERC721Metadata.sol";

// SPDX-License-Identifier: MIT
// OpenZeppelin Contracts (last updated v4.7.0) (token/ERC721/IERC721.sol)


contract NFT {
     // nft testnet https://nikau.uncoverexplorer.com/extrinsic/4750448-1 collection id - 261 and series id - 0
     // 0xff
     address nft = 0xaAAAaaAA00000000000000000000000000000000;


    function balanceOfProxy(address who) public view returns (uint256) {
        return IERC721(nft).balanceOf(who);
    }

    function ownerOfProxy(uint256 tokenId) public view returns (address) {
        return IERC721(nft).ownerOf(tokenId);
    }

    function nameOfNFT() public view returns (string memory) {
        return IERC721Metadata(nft).name();
    }

    function symbolOfNFT() public view returns (string memory) {
        return IERC721Metadata(nft).symbol();
    }

    function tokenURIOfNFT(uint256 serial_number) public view returns (string memory) {
        return IERC721Metadata(nft).tokenURI(serial_number);
    }
    
    
    function transferFromProxy(
        uint256 serial_number
    ) external {
         IERC721(nft).transferFrom(address(this), msg.sender, serial_number);
    }

     function approveFromProxy(
        uint256 serial_number
    ) external {
         IERC721(nft).approve(address(this), serial_number);
    }
}

```
Able to fetch all the state/read only information. However transaction fails
Similar contract for ERC20 works fine

```
pragma solidity ^0.8.13;
import "@openzeppelin/contracts/interfaces/IERC20.sol";

contract abc {
// CENNZ testnet address with assetId 16000
address cennz = 0xcCccccCc00003E80000000000000000000000000;

function balanceOfProxy(address who) public view returns (uint256) {
    return IERC20(cennz).balanceOf(who);
}

function transferFromProxy(uint256 amount) external {
    IERC20(cennz).transferFrom(msg.sender, address(this), amount);
}
}
```

**Comment by jordy25519** [comment]
can you provide some more info about the error/failure messages? the order of transactions attempted etc.

**Comment by KarishmaBothara** [comment]
<img width="506" alt="Screen Shot 2022-08-18 at 11 26 13 AM" src="https://user-images.githubusercontent.com/29415595/185260669-5e871f2c-d6b4-4031-bb59-7801b0360eed.png">

I can also explain the flow in code..

1. It goes from [runner.rs](https://github.com/cennznet/cennznet/blob/729105168699dd497b072c4d2a08f883be2af3a7/runtime/src/runner.rs#L139) call

2.  Gets to this library cargo/git/checkouts/evm-31951a45719dc0d6/52d5fe5/src/executor/stack/executor.rs
 `#[allow(clippy::too_many_arguments)]

	fn call_inner(
		&mut self,
		code_address: H160,
		transfer: Option<Transfer>,
		input: Vec<u8>,
		target_gas: Option<u64>,
		is_static: bool,
		take_l64: bool,
		take_stipend: bool,
		context: Context,
	) -> Capture<(ExitReason`
	
	The main line of interest in this function is 
	`let mut runtime = Runtime::new(Rc::new(code), Rc::new(input), context, self.config);
	let reason = self.execute(&mut runtime);
	`
	which internally tries to find opcode, I expect it to resolve with opcode
		/// `CALL`
	pub const CALL: Opcode = Opcode(0xf1);
	
	
![image](https://user-images.githubusercontent.com/29415595/185261817-53e23053-be3c-45cd-a5bc-7997b503cf03.png)

	
However it fails to get the right opcode and the routing_prefix does not matches the address of nft which is 0xaAAAaaAA00000000000000000000000000000000
	
Order of transaction attempted
1. Create NFT with eth wallet address, set series name.
2. Deploy the NFT contract above.
3. Fetch the state info successfully (name, symbol, tokenURI)
4. Try doing transaction in smart contract (tranferFrom, approve), both fails with above error

**Comment by KarishmaBothara** [comment]
https://hackmd.io/kSCrdeZKSE2L6SF2c2-69Q investigation so far

**Comment by ken-futureverse** [comment]
closes this in favour of #712
